File: test/Models/energybalance_pkg/src/record/Penman.cpp
// @@tagdynamic@@
// @@tagdepends: vle.discrete-time @@endtagdepends

#include <vle/DiscreteTime.hpp>
namespace vd = vle::devs;
namespace vv = vle::value;
// Definition du namespace de la classe du modele
namespace record {
namespace Energybalance {
using namespace vle::discrete_time;
class Penman: public DiscreteTimeDyn {
public:
    Penman(const vd::DynamicsInit& atom, const vd::InitEventList& events) : DiscreteTimeDyn(atom, events)
    {
        // Ces parametres ont une valeur par defaut utilisee si la condition n'est pas definie
        //...
        rhoDensityAir = (events.exist("rhoDensityAir")) ? vv::toDouble(events.get("rhoDensityAir")) : 1.225;

    }
    /**
    * @brief Destructeur de la classe du modèle.
    **/
    virtual ~Penman() {};
    /**
    * @brief Methode de calcul effectuée à chaque pas de temps
    * @param time la date du pas de temps courant
    */
    virtual void compute(const vd::Time& /*time*/)
    {
        evapoTranspirationPenman = evapoTranspirationPriestlyTaylor() / Alpha + (1000.0d * (rhoDensityAir * specificHeatCapacityAir * VPDair() * conductance() / (lambdaV * (hslope() + psychrometricConstant))));
    }
private:
    //Variables d'etat
    /**
    * @brief  evapoTranspiration of Penman Monteith (g/m**2/d)
    */
    Var evapoTranspirationPenman

    //Entrées
    /**
        * @brief conductance (m/d)
        */
    Var conductance/**
        * @brief evapoTranspiration of Priestly Taylor  ((g/m**2)/d)
        */
    Var evapoTranspirationPriestlyTaylor/**
        * @brief the slope of saturated vapor pressure temperature curve at a given temperature  (hPa/degC)
        */
    Var hslope/**
        * @brief  vapour pressure density (hPa)
        */
    Var VPDair

    //Paramètres du modele
    /**
    * @brief psychrometric constant (hPa/degC)
    */
    double psychrometricConstant;
    /**
    * @brief Priestley-Taylor evapotranspiration proportionality constant ()
    */
    double Alpha;
    /**
    * @brief latent heat of vaporization of water (MJ/kg)
    */
    double lambdaV;
    /**
    * @brief Density of air (kg/m**3)
    */
    double rhoDensityAir;
    /**
    * @brief Specific heat capacity of dry air ((MJ/kg)/degC)
    */
    double specificHeatCapacityAir;
};
}
}
DECLARE_DYNAMICS(record::Energybalance::Penman); // balise specifique VLE

File: src/pycropml/transpiler/antlr_py/tests/examples/openalea/SQ_Energy_Balance/Penman.py
# coding: utf8
from copy import copy
from array import array
from math import *
from typing import *
from datetime import datetime

import numpy

#%%CyML Model Begin%%
def model_penman(evapoTranspirationPriestlyTaylor:float,
         hslope:float,
         VPDair:float,
         psychrometricConstant:float,
         Alpha:float,
         lambdaV:float,
         rhoDensityAir:float,
         specificHeatCapacityAir:float,
         conductance:float):
    """
     - Name: Penman -Version: 1.0, -Time step: 1
     - Description:
                 * Title: Penman Model
                 * Authors: Peter D. Jamieson, Glen S. Francis, Derick R. Wilson, Robert J. Martin
                 * Reference:  https://doi.org/10.1016/0168-1923(94)02214-5
                 * Institution: New Zealand Institute for Crop and Food Research Ltd.,
                 New Zealand Institute for Crop and Food Research Ltd.,
                 New Zealand Institute for Crop and Food Research Ltd.,
                 New Zealand Institute for Crop and Food Research Ltd.
             
                 * ExtendedDescription: It uses Penmann-Monteith method vase on the availability of wind and vapor pressure daily data
                 * ShortDescription: It uses Penmann-Monteith method vase on the availability of wind and vapor pressure daily data
     - inputs:
                 * name: evapoTranspirationPriestlyTaylor
                               ** description : evapoTranspiration of Priestly Taylor 
                               ** variablecategory : rate
                               ** datatype : DOUBLE
                               ** default : 449.367
                               ** min : 0
                               ** max : 10000
                               ** unit : g m-2 d-1
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : variable
                 * name: hslope
                               ** description : the slope of saturated vapor pressure temperature curve at a given temperature 
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** default : 0.584
                               ** min : 0
                               ** max : 1000
                               ** unit : hPa °C-1
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : variable
                 * name: VPDair
                               ** description :  vapour pressure density
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** default : 2.19
                               ** min : 0
                               ** max : 1000
                               ** unit : hPa
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : variable
                 * name: psychrometricConstant
                               ** description : psychrometric constant
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** default : 0.66
                               ** min : 0
                               ** max : 1
                               ** unit : 
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : parameter
                 * name: Alpha
                               ** description : Priestley-Taylor evapotranspiration proportionality constant
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** default : 1.5
                               ** min : 0
                               ** max : 100
                               ** unit : 
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : parameter
                 * name: lambdaV
                               ** description : latent heat of vaporization of water
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** default : 2.454
                               ** min : 0
                               ** max : 10
                               ** unit : 
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : parameter
                 * name: rhoDensityAir
                               ** description : Density of air
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** default : 1.225
                               ** unit : 
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : parameter
                 * name: specificHeatCapacityAir
                               ** description : Specific heat capacity of dry air
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** default : 0.00101
                               ** min : 0
                               ** max : 1
                               ** unit : 
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : parameter
                 * name: conductance
                               ** description : conductance
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** min : 0
                               ** max : 10000
                               ** default : 598.685
                               ** unit : m d-1
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
                               ** inputtype : variable
     - outputs:
                 * name: evapoTranspirationPenman
                               ** description :  evapoTranspiration of Penman Monteith
                               ** variablecategory : rate
                               ** datatype : DOUBLE
                               ** min : 0
                               ** max : 5000
                               ** unit : g m-2 d-1
                               ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    """

    evapoTranspirationPenman:float
    evapoTranspirationPenman = evapoTranspirationPriestlyTaylor / Alpha + (1000.0 * (rhoDensityAir * specificHeatCapacityAir * VPDair * conductance / (lambdaV * (hslope + psychrometricConstant))))
    return evapoTranspirationPenman
#%%CyML Model End%%